[JSC] Fix Atomics.waitAsync by using weak references in JSC::DeferredWorkTimer

hyjorc1 opened the pull request:
#### 1e4577ba8de434bc8942b01baf6b1d65ba967ab1
<pre>
[JSC] Fix Atomics.waitAsync by using weak references in JSC::DeferredWorkTimer
<a href="https://bugs.webkit.org/show_bug.cgi?id=272896">https://bugs.webkit.org/show_bug.cgi?id=272896</a>
<a href="https://rdar.apple.com/126690921">rdar://126690921</a>

Reviewed by Yusuke Suzuki.

Previously, we introduced JSC::DeferredWorkTimer by keeping the task related
dependencies and environment alive with strong references. This works fine
until Atomics.waitAsync is introduced. Consider a scenario that an Atomics.waitAsync
registered in the main thread with infinity timeout is never be resolved in the
runtime. At some point, the corresponding JSGlobalObject is no longer needed.
But the JSGlobalObject cannot be collected due to the previous registered
Atomics.waitAsync&apos;s JSPromise is strongly referenced and queued in the
DeferredWorkTimer.

To fix this issue, we can transfer the liveness responsibility from the
DeferredWorkTimer::TicketData to the target&apos;s JSGlobalObject. And this is
how it works. Given the graph as shown below:

            VM &lt;----------------------------
            ^                              |
            |                              |
        JSGlobalObject_0      ----&gt; JSGlobalObject_n -----
            ^                 |            ^             |
            |                 |            |             |
SharedArrayBufferContents -----            | S           | S
            ^                              |             |
            |                       W      |             |
            Waiter ----&gt; TicketData -----&gt; Target &lt;------|
                            |                            |
                            |  W                         |
                            ------&gt; Dependencies &lt;--------

1. The JSGlobalObject_0 and JSGlobalObject_n may or may not be the same object,
   which is dependent on the VM.
2. The SharedArrayBufferContents is dependent on the multiple JSGlobalObjects.
3. The Waiter is dependent on the SharedArrayBufferContents.
4. The Waiter has a TicketData used to schedule a task to DeferredWorkTimer.
5. The TicketData replies on a Target and its corresponding Dependencies to complete the task.
6. The TicketData has weak references to the Target and Dependencies.
7. The Target&apos;s JSGlobalObject_n holds the strong references to the Target and Dependencies.

Then, we should do:
A. If JSGlobalObject_n is alive, then it should clear those strong references once the
   Waiter done the task.
B. If JSGlobalObject_n is dead, then it should clear the corresponding waiters first and
   then clear the other TicketDatas associating to it.

If no others are referencing JSGlobalObject_n and Target, GC would consider the strong reference
cycle in Target -&gt; JSGlobalObject_n -&gt; Target as dead status. Then, they are available
for recycling.

* Source/JavaScriptCore/heap/WeakInlines.h:
(JSC::operator==):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
(JSC::DeferredWorkTimer::TicketData::TicketData):
(JSC::DeferredWorkTimer::addPendingWork):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::addObjectsForDeferredTaskTimer):
* Source/JavaScriptCore/wasm/WasmStreamingCompiler.cpp:
(JSC::Wasm::StreamingCompiler::StreamingCompiler):
* Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
(JSC::JSWebAssembly::webAssemblyModuleValidateAsync):
(JSC::instantiate):
(JSC::compileAndInstantiate):

Canonical link: <a href="https://commits.webkit.org/279852@main">https://commits.webkit.org/279852@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9209306fc8ffc8a97dd1d13654ed6dccf8735dad

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54700 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34125 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7276 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57979 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5432 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41672 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/5442 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44308 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3669 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56795 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32260 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47384 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25432 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29053 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/4720 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3572 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/48065 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50788 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4933 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59569 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/54205 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29943 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5080 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51731 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31086 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47469 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51121 "Found 2 new API test failures: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/text/state-changed, /WebKitGTK/TestInspectorServer:/webkit/WebKitWebInspectorServer/http-test-page-list (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12000 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32079 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/66502 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30872 "Build is in progress. Recent messages:OS: Sonoma (14.5), Xcode: 15.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12674 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -2633,6 +2634,9 @@ void JSGlobalObject::visitChildrenImpl(JSCell* cell, Visitor& visitor)
     thisObject->m_regExpGlobalData.visitAggregate(visitor);
+
+    for (auto& object : thisObject->m_objectsForDeferredTaskTimer)
+        visitor.append(object);
```
Let's take a cellLock in this scope. GC marker can run concurrently.

kmiller68 replied:
How do these references ever get cleared?

hyjorc1 replied:
Once the JSGlobalObject is no longer needed by others, then only this remains. So next time when doing GC, they will be considered as that they are not referenced and then collected.

```
JSGlobalObject -> JSObjectsForDeferredTaskTimer
        ^                 |
        |                 |
        -------------------
```

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.h:
```diff
@@ -505,6 +505,13 @@ class JSGlobalObject : public JSSegmentedVariableObject {
+    void addObjectsForDeferredTaskTimer(JSObject* object)
+    {
+        m_objectsForDeferredTaskTimer.append(WriteBarrier<JSObject>());
+        m_objectsForDeferredTaskTimer.last().set(*m_vm, this, object);
```
Ditto. Let's take a cellLock.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -3268,4 +3282,33 @@ void JSGlobalObject::setWrapperMap(std::unique_ptr<WrapperMap>&& map)
+        value.at(i++).set(vm(), this, dependency.get());
+    value.at(i).set(vm(), this, scriptExecutionOwner);
+
+    m_objectsForTicket->add(ticket, WTFMove(value));
```
Let's first add values to the map, and set cells to that to ensure that emitting write-barrier after chaining.
```
auto addResult = m_objectsForTicket->add(ticket, WTFMove(value));
addResult->value.at(I++) = ...
````

hyjorc1 replied:
Thanks, updated.

Constellation commented on Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
```diff
@@ -3268,4 +3282,33 @@ void JSGlobalObject::setWrapperMap(std::unique_ptr<WrapperMap>&& map)
+
+    m_objectsForTicket->add(ticket, WTFMove(value));
+}
+void JSGlobalObject::removeObjectsForTicket(DeferredWorkTimer::Ticket ticket)
+{
+    Locker locker { cellLock() };
+    m_objectsForTicket->remove(ticket);
+}
+void JSGlobalObject::clearObjectsForTicket()
+{
+    if (!m_objectsForTicket)
+        return;
+
+    WaiterListManager::singleton().unregister(this);
+    // Clear the rest tickets safely.
+    vm().deferredWorkTimer->cancelPendingWorkSafe(this);
+}
```
Does it correctly work even though JSGlobalObject destruction can be delayed (due to incremental sweeper)?

hyjorc1 replied:
I think so, let's enumerate all cases here:

```
id JSGlobalObject::clearObjectsForTicket()
{
    if (!m_objectsForTicket)
        return;

    // Case 1: Before unregistering waiters in the list.

    WaiterListManager::singleton().unregister(this);      // <-- Step [1]

    // Case 2: After unregistering waiters in the list but before clearing the rest tickets.

    vm().deferredWorkTimer->cancelPendingWorkSafe(this);  // <-- Step [2]

    // Case 3: After all clearance is done.
}
```

Note that:
* Step [1] is guarded by the waiter list lock.
* Step [2] is guarded by the task lock in the timer.

Case 1: Before Step [1], if the fired timer (with a waiter in the list) acquired the list lock and scheduled the task, then the waiter was already cleared by WaiterListManager.
Case 1.1: If the DeferredWorkTimer got the task lock before Step [2], then it would finish the task safely.
Case 1.2: If Step [2] got the task lock, then the task would be cancelled. Then, the DeferredWorkTimer wouldn't do the task.

Case 2: Before Step [1], if the fired timer (with a waiter not in the list) scheduled the task, then it only needs to worry about Step [2].
Case 2.1: If the DeferredWorkTimer got the task lock before Step [2], then it would finish the task safely.
Case 2.2: If Step [2] got the task lock, then the task would be cancelled. Then, the DeferredWorkTimer wouldn't do the task.

Case 3: After Step [2], the rest tickets are cancelled and the empty tasks are scheduled for removing those tickets in the DeferredWorkTimer. Let's say a fired timer (with a waiter not in the list) scheduled a task for resolving the promise task.
Case 3.1: If the DeferredWorkTimer was doing the promise task first, then it would find the ticket is cancelled then stop working on the task. Next, the following empty task would remove the ticket completely.
Case 3.2: If the DeferredWorkTimer was doing the empty task frist, then the ticket would be removed completely. Next, the following promise task couldn't find the `ticket` in the `m_pendingTickets` since the `ticket` should be a nullptr at that time.
Case 3.3: If the empty task was done before the timer was fired, then just destuct the waiter itself only.

```
void DeferredWorkTimer::doWork(VM& vm)
{
    ...
    while (!m_tasks.isEmpty()) {
        auto [ticket, task] = m_tasks.takeFirst();
        dataLogLnIf(DeferredWorkTimerInternal::verbose, "Doing work on: ", RawPointer(ticket));

        auto pendingTicket = m_pendingTickets.find(ticket);
        // We may have already canceled this task or its owner may have been canceled.
        if (pendingTicket == m_pendingTickets.end())
            continue;
        ...
```

Anyway I added another check in `DeferredWorkTimer::TicketData::isCancelled()` for any unexpected cases.
```
bool isCancelled() const { return !scriptExecutionOwner.get() || !globalObject.get(); }
```

Constellation reviewed (commented):
I think we need to check that `Weak<JSGlobalObject>` is still alive in timer.

Constellation commented on Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
```diff
@@ -50,27 +48,31 @@ class DeferredWorkTimer final : public JSRunLoopTimer {
     private:
         WTF_MAKE_TZONE_ALLOCATED(TicketData);
```
Let's also annotate `TicketData` with `WTF_MAKE_NONCOPYABLE(TicketData)` etc. to ensure that this is not copyable.

Constellation reviewed (approved):
r=me

hyjorc1 committed:
[JSC] Fix Atomics.waitAsync by using weak references in JSC::DeferredWorkTimer
https://bugs.webkit.org/show_bug.cgi?id=272896
rdar://126690921

Reviewed by Yusuke Suzuki.

Previously, we introduced JSC::DeferredWorkTimer by keeping the task related
dependencies and environment alive with strong references. This works fine
until Atomics.waitAsync is introduced. Consider a scenario that an Atomics.waitAsync
registered in the main thread with infinity timeout is never be resolved in the
runtime. At some point, the corresponding JSGlobalObject is no longer needed.
But the JSGlobalObject cannot be collected due to the previous registered
Atomics.waitAsync's JSPromise is strongly referenced and queued in the
DeferredWorkTimer.

To fix this issue, we can transfer the liveness responsibility from the
DeferredWorkTimer::TicketData to the target's JSGlobalObject. And this is
how it works. Given the graph as shown below:

            VM <----------------------------
            ^                              |
            |                              |
        JSGlobalObject_0      ----> JSGlobalObject_n -----
            ^                 |            ^             |
            |                 |            |             |
SharedArrayBufferContents -----            | S           | S
            ^                              |             |
            |                       W      |             |
            Waiter ----> TicketData -----> Target <------|
                            |                            |
                            |  W                         |
                            ------> Dependencies <--------

1. The JSGlobalObject_0 and JSGlobalObject_n may or may not be the same object,
   which is dependent on the VM.
2. The SharedArrayBufferContents is dependent on the multiple JSGlobalObjects.
3. The Waiter is dependent on the SharedArrayBufferContents.
4. The Waiter has a TicketData used to schedule a task to DeferredWorkTimer.
5. The TicketData replies on a Target and its corresponding Dependencies to complete the task.
6. The TicketData has weak references to the Target and Dependencies.
7. The Target's JSGlobalObject_n holds the strong references to the Target and Dependencies.

Then, we should do:
A. If JSGlobalObject_n is alive, then it should clear those strong references once the
   Waiter done the task.
B. If JSGlobalObject_n is dead, then it should clear the corresponding waiters first and
   then clear the other TicketDatas associating to it.

If no others are referencing JSGlobalObject_n and Target, GC would consider the strong reference
cycle in Target -> JSGlobalObject_n -> Target as dead status. Then, they are available
for recycling.

* Source/JavaScriptCore/heap/WeakInlines.h:
(JSC::operator==):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.cpp:
(JSC::DeferredWorkTimer::TicketData::TicketData):
(JSC::DeferredWorkTimer::addPendingWork):
* Source/JavaScriptCore/runtime/DeferredWorkTimer.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::addObjectsForDeferredTaskTimer):
* Source/JavaScriptCore/wasm/WasmStreamingCompiler.cpp:
(JSC::Wasm::StreamingCompiler::StreamingCompiler):
* Source/JavaScriptCore/wasm/js/JSWebAssembly.cpp:
(JSC::JSWebAssembly::webAssemblyModuleValidateAsync):
(JSC::instantiate):
(JSC::compileAndInstantiate):

Canonical link: https://commits.webkit.org/279852@main
